### DISCUSSION TITLE
Container getLocalBounds performance issues

### DISCUSSION THREAD

**Discussion by BioRyajenka** [korlibs/korge#1476] [General]
Hi, 
`Container.getLocalBoundsInternal()` is heavy because it needs to iterate all children. It becomes bottleneck when you use `View.width`/`View.height` properties too often, and you usually want to use them often. So for each width/height it calculates bounds from scratch.

My particular problem is with UIScrollable. During update phase `getLocalBoundsOptimized()` is being called 4 times (indirectly), which gives poor performance. I will create PR to fix that for UIScrollable soon, but here I wanted to discuss ways to fix the source issue.

E.g. the most efficient way will be to cache the bounds and only recalculate them when necessary. E.g. by adding `dirty` flag. But it won't work because you can't track all your children position/size changes automatically. 

What is your thoughts? Does it worth efforts at all, or we have to live with it, always keeping in mind width/height props are slow?

**Comment by soywiz** [comment]
Im fine with propagating up the dirtyness of children tranforms. I believe that works already on UIView? At least there is a relayout somewhere. We might want to generalize it. Do you want to implement it yourself? Or your PR wont be related ro that?

**Reply by BioRyajenka** [reply]
For now I am focused on optimising UIScrollable for high amount of elements (even 200 elements have a significant impact on performance rn), so I will be happy if someone will catch it up.

How do you see it? Naive approach would be to iterate each ancestor and make it dirty. But it's not efficient and probably will slow everything down, no?

Another option: propagate dirtiness during update phase. Much more efficient but getLocalBounds will remain dirty until next update tick, which will confuse often.

One more option: cache bounds per frame. I.e. at the start of the frame all bounds considered dirty, but in scope on one frame they won't be re-calculated. Well, again, this solution also have drawbacks.

**Reply by soywiz** [reply]
IIRC it should be pretty fast:

If you are worried about iterating the ancesors everytime, it is usually not the case. You can do it immediately when the transform, x, y, width, height is changed, no need to do special stuff.
The trick here is: go to the parent if it is not dirty, make it dirty and go to the grandparent recursively. But, once one parent or ancestor is already dirty, you stop iterating. Because some other sibling, nephew or descendant have alread make it dirty. So if you when changing a lot of nodes, ancestors are updated only once.
There is a special case to take into account, and it is when changing the parent of a node / move the node (that is not really frequent in typical scenarios), in that case you can unconditionally set dirtiness just in case an edge case was done.

**Reply by BioRyajenka** [reply]
You're right, it sounds better than I thought. I hope I will have time to prepare PR this week

**Comment by soywiz** [comment]
Fixed here: https://github.com/korlibs/korge/pull/1487
